pandas/tests/reshape/merge/test_join.py

```diff
@@ -956,3 +956,35 @@ def test_join_empty(left_empty, how, exp):
             expected = expected.rename_axis("A")
 
     tm.assert_frame_equal(result, expected)
+
+
+def test_join_multiindex_categorical_output_index_dtype():
+    # GH#50906
+    df1 = DataFrame(
+        {
+            "idx1": Categorical(["a", "a", "a"]),
+            "idx2": Categorical(["a", "a", "b"]),
+            "data": [1, 2, 3],
+        }
+    ).set_index(["idx1", "idx2"])
+
+    df2 = DataFrame(
+        {
+            "idx1": Categorical(["a", "a", "a"]),
+            "idx2": Categorical(["a", "b", "b"]),
+            "data2": [1, 2, 3],
+        }
+    ).set_index(["idx1", "idx2"])
+
+    expected = DataFrame(
+        {
+            "idx1": Categorical(["a", "a", "a", "a"]),
+            "idx2": Categorical(["a", "a", "b", "b"]),
+            "data": [1, 2, 3, 3],
+            "data2": [1, 1, 2, 3],
+        }
+    ).set_index(["idx1", "idx2"])
+
+    for how in ["inner", "outer", "left", "right"]:
```

Comment:
Could you use `pytest.mark.parameterize` for this?

Reply:
Yes, of course. Added parametrize to this [commit](https://github.com/pandas-dev/pandas/pull/51088/commits/ea7fe922a841b60e9e706f2c4467437c4fb0e180)